File: Two Pointers/Fixed window/Minimum Recolors to Get K Consecutive Black Blocks/Intutive_approach.cpp
#include<bits/stdc++.h>
using namespace std;

//Time complexity :- O(N)
//Space complexity :- O(1)

//Approach :-
// Consider each window of size K, we can find the minimum number of white clored blocks in the window
// we can use a sliding window to solve this problem and the min white colored window will be our answer

//link :- https://leetcode.com/problems/minimum-recolors-to-get-k-consecutive-black-blocks/

class Solution {
public:
    int minimumRecolors(string blocks, int k) {
        int n = blocks.size(), j=0, whiteCnt = 0, res = k;

        while(j<k){
            if(blocks[j] == 'W')
                whiteCnt++;
            
            ++j;
        }
        res = whiteCnt;

        for(int i=0; i<(n-k); ++i){
            if(blocks[i] == 'W')
                whiteCnt--;
            
            if(blocks[j] == 'W')
                whiteCnt++;
            
            ++j;
            res = min(res, whiteCnt);
        }
        return res;
    }
};

File: Dynamic Programming/Kadane Algo/Maximum Product Subarray/Intutive_approach.ts
#include<bits/stdc++.h>
using namespace std;

// Time Complexity: O(n)
// Space Complexity: O(1)

// Approach :- 
// Here the some observation for this problem 
// 1. When all +ve then product of array will be the answer
// 2. when odd number of -ve element then either prefix before -ve element or after it will have the max product
// 3. when even number of -ve element then all the array product is answer 
// 4. whenever zeros comes, the array should be divided such we find answer in different subarrays splitted by 0s
// Hence, we'll find the product of each suffix and prefix element and maintain a max product for our answer
// this approach also shares the same concept as kadanes algo

// Link :- https://leetcode.com/problems/maximum-product-subarray/

function maxProduct(nums: number[]): number {
    let res: number = -Infinity;
    let n: number = nums.length;
    let prefix: number = 1, suffix: number = 1; 

    for(let i: number = 0; i<n; ++i){
        prefix = ((prefix === 0) ? 1 : prefix)*nums[i];
        suffix = ((suffix === 0) ? 1 : suffix)*nums[n-i-1];
        res = Math.max(res, prefix, suffix);
    }
    return res;
};